src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,530 @@
+                    // but we still need to persist the finished state.
+                    // We currently persist it on the second barrier here rather than first.
+                    // This is because we can't update state table in first epoch,
+                    // since it expects to have been initialized in previous epoch
+                    // (there's no epoch before the first epoch).
+                    if is_snapshot_empty {
+                        let finished_state = construct_initial_finished_state(pk_in_output_indices.len());
+                        for vnode in upstream_table.vnodes().iter_vnodes() {
+                            backfill_state.update_progress(vnode, BackfillProgressPerVnode::InProgress(finished_state.clone()));
+                        }
+                    }
+
+                    persist_state_per_vnode(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        false,
+                        &mut backfill_state,
+                        &mut committed_progress,
+                        &mut temporary_state,
+                    ).await?;
+
+                    self.progress.finish(barrier.epoch.curr);
+                    yield msg;
+                    break;
+                }
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                if let Message::Barrier(barrier) = &msg {
+                    self.state_table.commit_no_data_expected(barrier.epoch);
+                }
+                yield msg;
+            }
+        }
+    }
+
+    /// Read snapshot per vnode.
+    /// These streams should be sorted in storage layer.
+    /// 1. Get row iterator / vnode.
+    /// 2. Merge it with `select_all`.
+    /// 3. Change it into a chunk iterator with `iter_chunks`.
+    /// This means it should fetch a row from each iterator to form a chunk.
+    ///
+    /// NOTE(kwannoel): We interleave at chunk per vnode level rather than rows.
+    /// This is so that we can compute `current_pos` once per chunk, since they correspond to 1
+    /// vnode.
+    ///
+    /// TODO(kwannoel): Support partially complete snapshot reads.
+    /// That will require the following changes:
+    /// Instead of returning stream chunk and vnode, we need to dispatch 3 diff messages:
+    /// 1. COMPLETE_VNODE(vnode): Current iterator is complete, in that case we need to handle it
+    ///    in arrangement backfill. We should not buffer updates for this vnode, and forward
+    ///    all messages.
+    /// 2. MESSAGE(CHUNK): Current iterator is not complete, in that case we
+    ///    need to buffer updates for this vnode.
+    /// 3. FINISHED: All iterators finished.
```

Comment:
Take note of this as well, I think implementation should still be correct without this, just inefficient.
Currently we do not support early finish for a vnode partition. We still wait for all vnode partitions to complete backfill, before finishing the progress.

Plan to do it in separate PR.